### PR TITLE
Add check for enable-auto-restarts config option

### PR DESCRIPTION
**Pull request by rgildein**
Add sanity checks for generating upgrade plan for application, where we can check all relevant pre-checks, before generating upgrade plan. Add _check_auto_restarts to verify that the `enable-auto-restarts` config option is disabled.

### CODE REVIEW

**Comment by rgildein** [comment]
I think that we should also move following functions to `upgrade_plan_sanity_checks`:
- `cou.apps.base.OpenStackApplication._verify_channel`
- `cou.apps.base.OpenStackApplication.current_os_release` raising `MismatchedOpenStackVersions`

File: cou/apps/base.py [hunk 1]
```diff
@@ -713,3 +717,40 @@ def _get_wait_step(self) -> PostUpgradeStep:
...
+
+        If the enable-auto-restart is not enabled, this check will raise an exception.
+
+        :raises HaltUpgradePlanGeneration: When enable-auto-restarts is not enabled.
+        """
+        if "enable-auto-restarts" not in self.config:
+            logger.debug("%s application has not enable-auto-restarts config option", self.name)
+            return
+
+        if self.config.get("enable-auto-restarts") is False:
```
Comment (rgildein) [inline cou/apps/base.py:732]:
```suggestion
        if self.config["enable-auto-restarts"].get("value") is False:
```

File: cou/apps/base.py [hunk 2]
```diff
@@ -713,3 +717,40 @@ def _get_wait_step(self) -> PostUpgradeStep:
...
+
+    def _check_application_target(self, target: OpenStackRelease) -> None:
+        """Check if application release is not lower or equal to target.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :raises HaltUpgradePlanGeneration: When the application halt the upgrade plan generation.
+        """
+        logger.debug(
+            "%s application current os_relase is %s and apt source is %s",
```
Comment (aieri) [inline cou/apps/base.py:746]:
```suggestion
            "%s application current os_release is %s and apt source is %s",
```

File: cou/apps/base.py [hunk 3]
```diff
@@ -713,3 +717,40 @@ def _get_wait_step(self) -> PostUpgradeStep:
...
+
+    def _check_auto_restarts(self) -> None:
+        """Check if enable-auto-restarts is enabled.
+
+        If the enable-auto-restart is not enabled, this check will raise an exception.
+
+        :raises HaltUpgradePlanGeneration: When enable-auto-restarts is not enabled.
+        """
+        if "enable-auto-restarts" not in self.config:
+            logger.debug("%s application has not enable-auto-restarts config option", self.name)
```
Comment (aieri) [inline cou/apps/base.py:729]:
```suggestion
            logger.debug("%s application does not have an enable-auto-restarts config option", self.name)
```

File: cou/apps/base.py [hunk 4]
```diff
@@ -713,3 +717,40 @@ def _get_wait_step(self) -> PostUpgradeStep:
...
+            return
+
+        if self.config["enable-auto-restarts"].get("value") is False:
+            raise HaltUpgradePlanGeneration(
+                f"It is not safe to continue upgrading '{self.name}' application when the "
+                "'enable-auto-restart' is disabled. Please enable it before running COU again."
```
Comment (aieri) [inline cou/apps/base.py:735]: I suggest rewording as follows:
```suggestion
                f"COU does not currently support upgrading applications that disable service restarts. "
                "Please enable charm option enable-auto-restart and rerun COU to upgrade the "
                "{self.name} application."
```

**Review by aieri — CHANGES_REQUESTED** [review]
Just some rewording requests

File: cou/apps/base.py [hunk 5]
```diff
@@ -713,3 +717,43 @@ def _get_wait_step(self) -> PostUpgradeStep:
...
+        :raises HaltUpgradePlanGeneration: When enable-auto-restarts is not enabled.
+        """
+        if "enable-auto-restarts" not in self.config:
+            logger.debug(
+                "%s application does not have an enable-auto-restarts config option", self.name
+            )
+            return
+
+        if self.config["enable-auto-restarts"].get("value") is False:
+            raise HaltUpgradePlanGeneration(
```
Comment (gabrielcocenza) [inline cou/apps/base.py:735]: IIRC, `HaltUpgradePlanGeneration` was created to be able to stop the plan generation of an app, but not stop the generation of other apps. Is this the intention here? I'm not sure if we should continue the plan generation in this case

Comment (rgildein) [reply cou/apps/base.py:735]: You are right, I need to change it to ApplicationError instead.

File: cou/apps/base.py [hunk 6]
```diff
@@ -395,6 +387,18 @@ def post_upgrade_steps(
...
+        """Run sanity checks before generating upgrade plan.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :raises ApplicationError: When enable-auto-restarts is not enabled.
+        :raises HaltUpgradePlanGeneration: When the application halt the upgrade plan generation.
+        """
+        self._check_application_target(target)
+        self._check_auto_restarts()
+        logger.info("%s application met all the necessary prerequisites", self.name)
```
Comment (gabrielcocenza) [inline cou/apps/base.py:400]: Non-blocker: I think this log it's hard to understand without a little more context.
```suggestion
        logger.info("%s application met all the necessary prerequisites to generate the upgrade plan", self.name)
```

File: tests/unit/apps/test_base.py [hunk 7]
```diff
@@ -281,3 +281,69 @@ def test_get_reached_expected_target_step(mock_workload_upgrade, units, model):
 
     app._get_reached_expected_target_step(target, units)
     mock_workload_upgrade.assert_has_calls(expected_calls)
+
+
```
Comment (gabrielcocenza) [inline tests/unit/apps/test_base.py:285]: Maybe it worth parameterize the config and have the following scenarios:

1. no config for `enable-auto-restarts` (already implemented)
2. `enable-auto-restarts` set to True

File: cou/apps/base.py [hunk 8]
```diff
@@ -713,3 +719,43 @@ def _get_wait_step(self) -> PostUpgradeStep:
...
+
+        if self.config["enable-auto-restarts"].get("value") is False:
+            raise ApplicationError(
+                "COU does not currently support upgrading applications that disable service "
+                "restarts. Please enable charm option enable-auto-restart and rerun COU to "
+                f"upgrade the {self.name} application."
+            )
+
+    def _check_application_target(self, target: OpenStackRelease) -> None:
+        """Check if application release is not lower or equal to target.
```
Comment (agileshaw) [inline cou/apps/base.py:744]:
```suggestion
        """Check if application release is not lower than or equal to target.
```

File: cou/apps/base.py [hunk 9]
```diff
@@ -713,3 +719,43 @@ def _get_wait_step(self) -> PostUpgradeStep:
             parallel=False,
             coro=self.model.wait_for_active_idle(self.wait_timeout, apps=apps),
         )
+
+    def _check_auto_restarts(self) -> None:
+        """Check if enable-auto-restarts is enabled.
+
+        If the enable-auto-restart is not enabled, this check will raise an exception.
```
Comment (agileshaw) [inline cou/apps/base.py:726]:
```suggestion
        If the enable-auto-restart option is not enabled, this check will raise an exception.
```

File: tests/unit/apps/test_subordinate.py [hunk 10]
```diff
@@ -282,15 +282,19 @@ def test_generate_plan_from_to(model, from_os, to_os):
...
     ],
 )
-def test_generate_plan_in_same_version(model, from_to):
-    """Test generate upgrade plan for OpenStackSubordinateApplication in same version."""
+def test_generate_plan_in_same_channel(model, from_to):
+    """Test generate upgrade plan for OpenStackSubordinateApplication in same channel.
+
+    The version based on apt_source_codename is ussuri because the app series is central
+    and current os_release is defined as targeted version. Because of this, we couldn't
+    use from_to=ussuri because it will raise HaltUpgradePlanGeneration.
```
Comment (agileshaw) [inline tests/unit/apps/test_subordinate.py:296]:
```suggestion
    The version based on apt_source_codename is ussuri because the app series is central
    and current os_release is defined as targeted version. Therefore using
    from_to=ussuri, which will raise HaltUpgradePlanGeneratio, is not possible.
```

File: cou/apps/base.py [hunk 11]
```diff
@@ -713,3 +719,43 @@ def _get_wait_step(self) -> PostUpgradeStep:
...
+        logger.debug(
+            "%s application current os_release is %s and apt source is %s",
+            self.name,
+            self.current_os_release,
+            self.apt_source_codename,
+        )
+
+        if self.current_os_release >= target and self.apt_source_codename >= target:
+            raise HaltUpgradePlanGeneration(
+                f"Application '{self.name}' already configured for release equal or greater "
```
Comment (agileshaw) [inline cou/apps/base.py:759]:
```suggestion
                f"Application '{self.name}' already configured for release equal to or greater "
```

**Review by agileshaw — CHANGES_REQUESTED** [review]
The approach LGTM. I just have some nit comments about wording

**Review by gabrielcocenza — APPROVED** [review]

**Review by aieri — APPROVED** [review]

**Review by agileshaw — APPROVED** [review]
